### PR TITLE
Add argument parsing for gzip.

### PR DESCRIPTION
**Pull request by krashanoff**
Pretty basic, but we need to get the ball rolling somewhere. This PR adds argument parsing and management for a new gzip using [clap's](https://clap.rs/) derive construct. Help messages have been recreated as doc comments.

The only notable change is that instead of handling input `FILE`s as strings, we interpret them as `PathBuf`s. I am open to discussion on whether this is of concern.

### CODE REVIEW

**Comment by TimothyGu** [comment]
> The only notable change is that instead of handling input `FILE`s as strings, we interpret them as `PathBuf`s. I am open to discussion on whether this is of concern.

I'm guessing this is only different on Windows (where UTF-16 is used for paths)? In that case, it would seem that `PathBuf` is the right thing to do.

File: src/main.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    #[clap(short, long)]
+    test: bool,
+
+    /// Compress faster
+    #[clap(short = '1', long)]
+    fast: bool,
+
+    /// Compress better
+    #[clap(short = '9', long)]
+    best: bool,
```
Comment (TimothyGu) [inline src/main.rs]: Maybe overkill for now, but GNU gzip also has support for `-2`…`-8` that we should support eventually. I wonder if clap has an easy way of supporting that.

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    #[clap(value_name = "FILE")]
+    files: Vec<PathBuf>,
+}
+
+fn main() {
+    let _args = Args::parse();
```
Comment (TimothyGu) [inline src/main.rs:188]: Do you have any idea how compatible `clap::Args` is with GNU getopt (or whatever library GNU gzip uses)? There are a lot of edge cases around how getopt works (`-c file` vs `file -c`, `-cd` vs `-c d` vs `-c -d`).

File: src/main.rs (hunk 1 above)
Comment (krashanoff) [reply src/main.rs]: I missed those options! I'll add them into this PR after I check out the getopt interop issue. We will also have to add `--rsyncable` and `--synchronous` (though Professor Eggert didn't seem super concerned about the latter).

File: src/main.rs (hunk 2 above)
Comment (krashanoff) [reply src/main.rs:188]: I'm not entirely sure, to be frank. I'll look into it later today. Worst comes to worst, we can roll our own like [exa](https://github.com/ogham/exa) did. I do think that using clap would be preferred, though, as it is the de facto argument parser in the Rust world.

File: src/main.rs [hunk 3]
```diff
@@ -67,17 +69,90 @@ struct Args {
...
+                if c == '=' {
+                    return;
+                }
+                if let Some(d) = c.to_digit(10) {
+                    level = Some(d as usize);
+                } else {
+                    cleaned.push(c);
+                }
+            });
+            if !["", "-"].contains(&cleaned.as_str()) {
```
Comment (TimothyGu) [inline src/main.rs]: We shouldn't merely ignore these strings. If we have `gzip ''` then the `''` should be treated as an empty file name. Similarly with `gzip -`, except the `-` gets treated as stdin.

Also, what if we have `--best -5`? I believe in general `-5` should override but that doesn't seem to happen right now. (Edit: I just saw that you wrote a comment about this issue.)

A better option (though a bit awkward) is to have `level2`…`level8` flags in the derived struct. This is also a more direct translation of the gzip C code which calls `getopt` with `123456789`. It also takes care of the `--best -5` problem earlier.

**Comment by krashanoff** [comment]
It's a bit strange, but clap doesn't support the ability to map multiple options to a single value in an argument struct deriving `Parser`. I opted to partition the arguments into the last occurring numerical flag and a set of strings, cleaned of said numerical flags, before passing control to clap.

With respect to clap's interoperability with `getopt` implementations, as far as my (basic) testing and review of gzip's implementation of option handling has concluded, it checks out.

The current gzip behavior simply uses the last-occurring compression quality specifier as the setting for the program. Since we have `--fast` and `--best` options distinct from numerical flags, this implementation will instead let the highest compression option specified to predominate. That is, if an user specifies `--best`, but also specifies `-3`, gzip will opt to compress using `--best`.

As this would technically be release 2.0 of gzip, I feel this behavior may be desirable as it is better defined. Should we want to recreate gzip's behavior to a T, we could manually check which option comes last by iterating through each argument. I worry this may be inching towards defeating the purpose of using clap, though.

One more point of weirdness I should address is that in order to invoke clap's generated error, we first convert it to an `App` ([L106](https://github.com/uclaacm/gzip/pull/1/commits/c5bae0f0c9095c0a39fdace83f8ceb5e8c4c35f8#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR106)).

**Comment by TimothyGu** [comment]
> I feel this behavior may be desirable as it is better defined

I'd actually prefer the original behavior. As an example, folks often set `CC='gcc -Wall'` etc. when compiling projects; and if the project explicitly want to make sure a warning is disabled, they can do `$CC -Wno-blah-blah` where later flags override earlier. This behavior is fairly common and well-defined, and I think we should respect the tradition here.

**Comment by krashanoff** [comment]
The GCC example really helped illustrate the issue. I think the final approach for this, then, will be to create `level2 ... level8` in our argument struct and lever the `overrides_with` constraint. TIL this overriding behavior is [POSIX-compliant](https://docs.rs/clap/latest/clap/struct.Arg.html#method.overrides_with)!

File: Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "gzip"
+version = "2.0.0"
```
Comment (TimothyGu) [inline Cargo.toml]: nit: put alpha or beta or some other indicator that this is not a release. Or in more old-school style, 1.90.0.

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+            "level_6",
+            "level_7",
+            "level_8",
+            "level_9"
+        ],
+    )]
+    level_9: bool,
+
+    /// Use suffix SUF on compressed files
+    #[clap(short = 'S', long, default_value = "")]
```
Comment (TimothyGu) [inline src/main.rs]:
```suggestion
    #[clap(short = 'S', long, default_value = "", value_name = "SUF")]
```

File: src/main.rs [hunk 6]
```diff
@@ -0,0 +1,254 @@
+use std::path::PathBuf;
+
+use clap::Parser;
+
+const DEFAULT_COMPRESSION_LEVEL: usize = 6;
```
Comment (TimothyGu) [inline src/main.rs]: Why usize, rather than something like int?

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            "level_3",
+            "level_4",
+            "level_5",
+            "level_6",
+            "level_7",
+            "level_8",
+            "level_9"
+        ],
+    )]
+    level_9: bool,
```
Comment (TimothyGu) [inline src/main.rs:152]: I liked the idea of having a separate integral `compression_level` field populated from the boolean flags (or the default). Hopefully not too much work to make one?

**Review by TimothyGu — APPROVED** [review]
LGTM, remaining are nits

File: src/main.rs [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    #[clap(short, long)]
+    no_name: bool,
+
+    /// Don't save or restore the original file time
+    #[clap(short = 'm')]
+    no_time: bool,
+
+    /// Save or restore the original modification time
+    #[clap(short = 'M', long)]
+    time: bool,
```
Comment (TimothyGu) [inline src/main.rs]: The help messages are guarded by `UNDOCUMENTED` in gzip.c. Let's do `hide = true` for now.

File: src/main.rs [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+            "level_1",
+            "level_2",
+            "level_3",
+            "level_4",
+            "level_5",
+            "level_6",
+            "level_7",
+            "level_8",
+            "level_9"
+        ],
```
Comment (TimothyGu) [inline src/main.rs]: I figured out that we could de-duplicate this by adding a static global variable:
```rust
static ALL_LEVEL_FLAGS: &'static [&'static str] = &[
    "level_1",
    "level_2",
    "level_3",
    "level_4",
    "level_5",
    "level_6",
    "level_7",
    "level_8",
    "level_9"
];
```
And then just say `overrides_with_all = ALL_LEVEL_FLAGS`

File: src/main.rs [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        hide = true
+    )]
+    level_8: bool,
+
+    /// Compress better
+    #[clap(
+        name = "level_9",
+        short = '9',
+        long = "best",
+        multiple_occurrences = true,
```
Comment (TimothyGu) [inline src/main.rs]: Why should this be true?

File: src/main.rs [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    #[clap(short, long)]
+    force: bool,
+
+    /// Keep (don't delete) input files
+    #[clap(short, long)]
+    keep: bool,
+
+    /// Don't save or restore the original file name
+    #[clap(short, long)]
+    no_name: bool,
```
Comment (TimothyGu) [inline src/main.rs:50]: We are missing a `-N, --name` boolean which should override this one.

File: src/main.rs [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+    /// Don't ask questions, compress links
+    #[clap(short, long)]
+    force: bool,
+
+    /// Keep (don't delete) input files
+    #[clap(short, long)]
+    keep: bool,
+
+    /// Don't save or restore the original file name
```
Comment (TimothyGu) [inline src/main.rs]:
```suggestion
    /// Don't save or restore the original file name and timestamp
```

according to gzip.c

**Comment by TimothyGu** [comment]
Also it's quite unfortunate that such a simple file is already 2.7M (on Linux x86-64)… (528K with `cargo rustc --release -- -C prefer-dynamic` + stripping).

File: src/main.rs (hunk 6 above)
Comment (krashanoff) [reply src/main.rs]: I had anticipated that whatever zlib bindings we chose would use `usize`. I think we've settled on [flate2](https://github.com/rust-lang/flate2-rs), though, so this should be a `u32`.

File: src/main.rs (hunk 7 above)
Comment (krashanoff) [reply src/main.rs:152]: While I was tinkering, the way I had it set was as a function `Args::compression_level(&self) -> u32`.

**Comment by krashanoff** [comment]
> Also it's quite unfortunate that such a simple file is already 2.7M (on Linux x86-64)… (528K with cargo rustc --release -- -C prefer-dynamic + stripping).

Yeah, I'm not too huge about this either.

```
❯ ls -al "$(which gzip)"
.rwxr-xr-x 97k root 13 Dec  2019 /usr/bin/gzip
```

To call this a bloat regression might be treating it lightly. I don't want to get too off-track here, but if there's one thing Rust doesn't have going for it, it's binary size.

Thank you for your input and review on this PR out of the blue, Tim. Hopefully we'll be able to get some more of your insights as the project continues!

**Pull request merged**